File: lattice.core/Throwables/LatticeException.vb
Namespace Throwables
    Public Class LatticeException
        Inherits Exception

        Public Property ErrorCode As String
        Public Property HelpText As String
        Public Property Location As String
        Public Property Notes As New List(Of String)()

        Public Sub New(message As String, Optional errorCode As String = "L000", Optional helpText As String = "", Optional location As String = "", Optional notes As IEnumerable(Of String) = Nothing)
            MyBase.New(message)
            Me.ErrorCode = errorCode
            Me.HelpText = helpText
            Me.Location = location
            If notes IsNot Nothing Then
                Me.Notes.AddRange(notes)
            End If
        End Sub

        Public Overrides ReadOnly Property Message As String
            Get
                Dim sb As New System.Text.StringBuilder()
                sb.AppendLine($"error[{ErrorCode}]: {MyBase.Message}")
                If Not String.IsNullOrEmpty(Location) Then
                    sb.AppendLine($"  --> {Location}")
                End If

                For Each n In Notes
                    sb.AppendLine($"  = note: {n}")
                Next

                If Not String.IsNullOrEmpty(HelpText) Then
                    sb.AppendLine()
                    sb.AppendLine($"  = help: {HelpText}")
                End If
                Return sb.ToString()
            End Get
        End Property
    End Class
End Namespace

File: Lattice.Runtime/Runtime/LatticeException.cs
namespace lattice.Throwables;

public class LatticeException : Exception
{
    public string ErrorCode { get; set; }
    public string HelpText { get; set; }
    public string Location { get; set; }
    public List<string> Notes { get; set; } = new();

    public LatticeException(string message, string errorCode = "L000", string helpText = "", string location = "", IEnumerable<string>? notes = null)
        : base(message)
    {
        ErrorCode = errorCode;
        HelpText = helpText;
        Location = location;
        if (notes != null) Notes.AddRange(notes);
    }

    public override string Message
    {
        get
        {
            var sb = new System.Text.StringBuilder();
            sb.AppendLine($"error[{ErrorCode}]: {base.Message}");
            if (!string.IsNullOrEmpty(Location))
                sb.AppendLine($"  --> {Location}");
            foreach (var n in Notes)
                sb.AppendLine($"  = note: {n}");
            if (!string.IsNullOrEmpty(HelpText))
            {
                sb.AppendLine();
                sb.AppendLine($"  = help: {HelpText}");
            }
            return sb.ToString().TrimEnd();
        }
    }
}
